[JENKINS-61046] - Allow plugins to force an update of the UpdateCenter

jtnord opened the pull request:
<!-- Comment: 
A great PR typically begins with the line below.
Replace XXXXX with the numeric part of the issue's id you created on JIRA.
Please note that if you want your changes backported into LTS, you will need to create a JIRA ticket for it. Read https://jenkins.io/download/lts/#backporting-process for more.
-->
See [JENKINS-61046](https://issues.jenkins-ci.org/browse/JENKINS-61046).

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* Developer: Allow plugins to force an update of an `UpdateSite`


<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [x] There are at least 2 approvals for the pull request and no outstanding requests for change
- [x] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [x] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [x] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

jtnord committed:
Allow plugins to force an update of the UpdateCenter

res0nance commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -180,7 +180,6 @@ public long getDataTimestamp() {
         }
     }
 
-    @Restricted(NoExternalUse.class)
```
Since this becomes part of the public API could you document it?

jtnord replied:
addressed in 9d642ebd8c08d8642ef3a433f1a4035d01a1d4f7

jtnord commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -180,7 +180,6 @@ public long getDataTimestamp() {
     }
 
-    @Restricted(NoExternalUse.class)
     public @Nonnull FormValidation updateDirectlyNow(boolean signatureCheck) throws IOException {
```
```suggestion
    /**
     * Forces an update the data file from the configured URL, irrespective of the last time the data was retrieved.
     * @param signatureCheck whether to enforce the signature (may be off only for testing!)
     * @return A {@code FormValidation} indicating the if the update metadata was successfully downloaded from the configured update site
     * @since TODO
     * @throws IOException if there was an error downloading or saving the file.
     */
    public @Nonnull FormValidation updateDirectlyNow(boolean signatureCheck) throws IOException {
```

jtnord committed:
Add javadoc to updateDirectlyNow

res0nance commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -180,7 +180,13 @@ public long getDataTimestamp() {
 
-    @Restricted(NoExternalUse.class)
+    /**
+     * Forces an update the data file from the configured URL, irrespective of the last time the data was retrieved.
```
```suggestion
     * Forces an update of the data file from the configured URL, irrespective of the last time the data was retrieved.
```

oleg-nenashev reviewed (changes requested):
Looks good to me in principle, but IMO we should avoid exposing test parameters in new API. I suggest to introduce a new API method which does not allow disabling the signature check. `(may be off only for testing!)` in Javadoc was a good first step, but IMHO it is not enough

timja commented:
> Looks good to me in principle, but IMO we should avoid exposing test parameters in new API. I suggest to introduce a new API method which does not allow disabling the signature check. `(may be off only for testing!)` in Javadoc was a good first step, but IMHO it is not enough

if we're introducing a new API then it would be good for it not to return form validation and either return a boolean or throw on error

jtnord committed:
[JENKINS-61046] introduce updateDirectlyNow() and updateDirectly()

@oleg-nenashev notes that the public API should not expose the security
option that is for testing.  Introduced updateDirectlyNow() as the new
API and updateDirectly() to have a mirror and deprecated
updateDirectly(boolean) as a consequence.

jtnord commented:
> if we're introducing a new API then it would be good for it not to return form validation and either return a boolean or throw on error

@timja that would collide with the way I have just addressed @oleg-nenashev's feedback.  internal non test code no longer call `updateDirectly(boolean)` or `updateDirectlyNow(boolean)` and changing the return would mean they would go back to calling the boolean parametered methods or a larger refactoring.

before I make any further changes could you both come to an agreement on the latest  or disagree on the merits of 20a2c768a164d7ee474e64cf419e495167a89cd3

jtnord committed:
fix javadoc grammar

fixes the grammar of the UpdateSite#updateDirectlyNow javadoc

Co-Authored-By: Raihaan Shouhell <raihaanhimself@gmail.com>

timja reviewed (approved):
It would be nice if an intended API didn’t return web specific form validation, but not blocking it

oleg-nenashev reviewed (approved):
Thanks @jtnord ! The current version looks good to me.

oleg-nenashev commented:
I suggest merging it tomorrow if no negative feedback

jtnord commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -160,14 +160,27 @@ public long getDataTimestamp() {
      * @since 1.502
+     * @deprecated use {@linkplain #updateDirectly()}
      */
+    @Deprecated
```
that would break source compatibility, happy to do that, but that could break some (at least internal) workflows, and goes against the Jenkins API deprecation policy.  Lets at least deprecate it for at least one LTS version first (could we then start a sweep of all deprecated API and look at restrict them?)

oleg-nenashev replied:
Fine with me. We break compatibility too often in recent releases
